lib/b3fs/src/bucket/dir/reader.rs

```diff
@@ -1,35 +1,347 @@
+use std::borrow::Borrow;
+use std::cell::Cell;
+use std::io::SeekFrom;
+use std::mem;
+use std::ops::Deref;
+
+use fastbloom_rs::{BloomFilter, Membership};
 use tokio::fs::{self};
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncSeekExt, BufReader};
+use tokio::sync::OnceCell;
 use triomphe::Arc;
 
 use super::iter::DirEntriesIter;
+use crate::bucket::dir::phf::{displace, hash, HasherState};
+use crate::bucket::errors;
 use crate::collections::HashTree;
-use crate::entry::BorrowedEntry;
+use crate::entry::{BorrowedEntry, BorrowedLink, OwnedLink};
+use crate::hasher::dir_hasher::B3_DIR_IS_SYM_LINK;
 
 pub struct B3Dir {
     /// The number of entries in this directory.
     num_entries: u16,
     /// The actual header file.
     file: Arc<fs::File>,
+    bloom_filter: OnceCell<BloomFilter>,
+    hashtree: OnceCell<HashTree<'static>>,
+    position_start_hashes: u64,
+    position_start_bloom_filter: u64,
+    position_start_entries: u64,
+    position_start_phf_table: u64,
+    phf_table: OnceCell<HasherState>,
+}
+
+#[macro_use]
+macro_rules! check_eof {
+    ($flag:expr) => {
+        if let Err(e) = $flag {
+            if e.kind() == std::io::ErrorKind::UnexpectedEof {
+                return Ok(None);
+            } else {
+                return Err(errors::ReadError::from(e));
+            }
+        }
+    };
 }
 
 impl B3Dir {
     pub(crate) fn new(num_entries: u32, file: Arc<fs::File>) -> Self {
-        debug_assert!(num_entries <= (u16::MAX as u32));
+        debug_assert!(num_entries <= u16::MAX as u32);
+        let position_start_hashes = 4 + 4 as u64;
+        let position_start_bloom_filter =
+            position_start_hashes + (num_entries as usize * 32) as u64;
```

Comment:
`num_entries` are the number of leafs in the hash tree. the total number of nodes would be `num_entries * 2 - 1`



```suggestion
            position_start_hashes + (num_entries as u64) * 64 - 32;
```